### PR TITLE
Delete `ByronToAllegraEra` eon

### PR DESCRIPTION
**Pull request by newhoggy**
# Changelog

```yaml
- description: |
    New `caseByronOrShelleyToAllegraOrMaryEraOnwards` function
    New `TxOutAdaOnlyByron` constructor
    Modify `TxOutAdaOnly` constructor to take `ShelleyToAllegraEra` instead of `ByronToAllegraEra`
    Delete `ByronToAllegraEra` eon
    Delete:
    - `caseByronToAllegraOrMaryEraOnwards`
    - `shelleyToAllegraEraToByronToAllegraEra`
    - `byronToAllegraEraConstraints`
    - `byronToAllegraEraToCardanoEra`
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

We desire to avoid eons that straddle the Byron/ShelleyBasedEra boundary.  Byron is very legacy so it ought to be a bit special and kept separate.

The reason we needed `ByronToAllegraEra` eon was because we had code that wanted to treat Byron era and subsequent eras the same way.  If we give Byron era it's own constructor, the need goes away.

# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-api/internal/Cardano/Api/TxBody.hs [hunk 1]
```diff
@@ -886,7 +886,9 @@ deriving instance Show (TxInsReference build era)
 
 data TxOutValue era where
 
-  TxOutAdaOnly :: ByronToAllegraEra era -> Lovelace -> TxOutValue era
```
Comment (newhoggy) [inline cardano-api/internal/Cardano/Api/TxBody.hs:889]: We split this constructor to avoid needing an extra eon.

Comment (carbolymer) [reply cardano-api/internal/Cardano/Api/TxBody.hs:889]: Not sure that's worth it. We're removing one eon, but the additional constructor is not handling any special behaviour and makes code a bit confusing.

**Review by carbolymer — COMMENTED** [review]
https://github.com/input-output-hk/cardano-api/pull/356/files#r1383315760
I'm not sure if this change is beneficial.

File: cardano-api/internal/Cardano/Api/TxBody.hs (hunk 1 above)
Comment (newhoggy) [reply cardano-api/internal/Cardano/Api/TxBody.hs:889]: Yeah, there's a tradeoff.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Let's refrain from more changes to the api concerning eons. We need to decide a way forward.

**Comment by newhoggy** [comment]
Closing for the moment.  Revisit later.

**Pull request closed without merging**